src/checker/nlprules.rs

```diff
@@ -0,0 +1,143 @@
+            || {
+                Ok(Tokenizer::from_reader(&mut &*DEFAULT_TOKENIZER_BYTES)
+                    .expect("build.rs pulls valid tokenizer description. qed"))
+            },
+            |path| -> Result<Tokenizer> {
+                let f = fs::File::open(&path)?;
+                Ok(Tokenizer::from_reader(f)?)
+            },
+        )?;
+
+        let rules = config.override_rules.as_ref().map_or_else(
+            || {
+                Ok(Rules::from_reader(&mut &*DEFAULT_RULES_BYTES)
+                    .expect("build.rs pulls valid tokenizer description. qed"))
+            },
+            |path| -> Result<Rules> {
+                let f = fs::File::open(&path)?;
+                Ok(Rules::from_reader(f)?)
+            },
+        )?;
+
+        let rules = rules
+            .into_iter()
+            .filter(|rule| {
+                match rule.category_id().to_lowercase().as_ref() {
+                    // The hunspell backend is aware of
+                    // custom lingo, which this one is not,
+                    // so there would be a lot of false
+                    // positives.
+                    "misspelling" => false,
+                    // Anything quotes related is not relevant
+                    // for code documentation.
+                    "typography" => false,
+                    _other => true,
+                }
+            })
+            .collect::<Rules>();
+        let rules = &rules;
+        let tokenizer = &tokenizer;
+        let suggestions = docu
+            .par_iter()
+            .try_fold::<SuggestionSet, Result<_>, _, _>(
+                || SuggestionSet::new(),
+                move |mut acc, (origin, chunks)| {
+                    debug!("Processing {}", origin.as_path().display());
+
+                    for chunk in chunks {
+                        acc.extend(
+                            origin.clone(),
+                            check_sentence(origin.clone(), chunk, tokenizer, rules),
+                        );
+                    }
+                    Ok(acc)
+                },
+            )
+            .try_reduce(
+                || SuggestionSet::new(),
+                |mut a, b| {
+                    a.join(b);
+                    Ok(a)
+                },
+            )?;
+
+        Ok(suggestions)
+    }
+}
+
+/// Check one segmented sentence
+fn check_sentence<'a>(
+    origin: ContentOrigin,
+    chunk: &'a CheckableChunk,
+    tokenizer: &Tokenizer,
+    rules: &Rules,
+) -> Vec<Suggestion<'a>> {
+    let plain = chunk.erase_cmark();
+    trace!("{:?}", &plain);
+    let txt = plain.as_str();
+
+    let mut acc = Vec::with_capacity(32);
+
+    let nlpfixes = rules.suggest(txt, tokenizer);
+    if nlpfixes.is_empty() {
+        return Vec::new();
+    }
+
+    'nlp: for NlpFix {
+        message,
+        start,
+        end,
+        replacements,
+        ..
+    } in nlpfixes
+    {
+        if start > end {
```

Comment:
Nlp actually produces fixes with such indices?

Reply:
I don't know, I just wanted to be safe rather than sorry.